sympy/combinatorics/pc_groups.py

```diff
@@ -0,0 +1,495 @@
+
+        Examples
+        ========
+        >>> from sympy.combinatorics.named_groups import SymmetricGroup
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> G = SymmetricGroup(3)
+        >>> PcGroup = G.polycyclic_group()
+        >>> collector = PcGroup.collector
+        >>> F, x0, x1 = free_group("x0, x1")
+        >>> w = x1*x0
+        >>> collector.map_relation(w)
+        x1**2
+
+        """
+        array = w.array_form
+        s1 = array[0][0]
+        s2 = array[1][0]
+        key = ((s2, -1), (s1, 1), (s2, 1))
+        key = self.free_group.dtype(key)
+        return self.pc_presentation[key]
+
+
+    def collected_word(self, word):
+        """
+        Return the collected form of a word.
+
+        A word `w` is called collected, if `w = x{i_1}**a_1*...*x{i_r}**a_r`
+        with `i_1 < i_2< ... < i_r` and `a_j` is in `{1, ..., s_j-1}`
+        if `s_j != infinity`.
+        Otherwise w is uncollected.
+
+        Examples
+        ========
+        >>> from sympy.combinatorics.named_groups import SymmetricGroup
+        >>> from sympy.combinatorics.perm_groups import PermutationGroup
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> G = SymmetricGroup(4)
+        >>> PcGroup = G.polycyclic_group()
+        >>> collector = PcGroup.collector
+        >>> F, x0, x1, x2, x3 = free_group("x0, x1, x2, x3")
+        >>> word = x3*x2*x1*x0
+        >>> collected_word = collector.collected_word(word)
+        >>> free_to_perm = {}
+        >>> free_group = collector.free_group
+        >>> for sym, gen in zip(free_group.symbols, collector.pcgs):
+        ...     free_to_perm[sym] = gen
+        >>> G1 = PermutationGroup()
+        >>> for w in word:
+        ...     sym = w[0]
+        ...     perm = free_to_perm[sym]
+        ...     G1 = PermutationGroup([perm] + G1.generators)
+        >>> G2 = PermutationGroup()
+        >>> for w in collected_word:
+        ...     sym = w[0]
+        ...     perm = free_to_perm[sym]
+        ...     G2 = PermutationGroup([perm] + G2.generators)
+        >>> G1 == G2
+        True
+
+        """
+        free_group = self.free_group
+        while True:
+            w = self.minimal_uncollected_subword(word)
+            if not w:
+                break
+
+            low, high = self.subword_index(word, free_group.dtype(w))
+            if low == -1:
+                continue
+            s1, e1 = w[0]
+            if len(w) == 2 and w[1][1] > 0:
+                s2, e2 = w[1]
+                s2 = ((s2, 1), )
+                s2 = free_group.dtype(s2)
+                word_ = self.map_relation(free_group.dtype(w))
+                word_ = s2*word_**e1
+                word_ = free_group.dtype(word_)
+                word = word.substituted_word(low, high, word_)
+
+            if not self.relative_order[self.index[s1]]:
+                continue
+            re = self.relative_order[self.index[s1]]
+            q = e1 // re
```

Comment:
Are these lines necessary? If all minimal uncollected subwords of length 1 are handled first, we should have `0 < e1 < re`.

Reply:
I think it will be `0 < e1 <= re`.

Reply:
`e1` should not be equal to `re` as `x1**e1` has already been collected using the power relation.

Reply:
`x1**e1` type will be collected from here, the above if was to handle the word made up of 2-free_groups.

Reply:
First if was for conjugate relators, this is for power relators.

Reply:
Hmm... It seems to me that this code will also be run in the conjugate relator case `len(w) == 2`. I think that it would be best to handle `len(w) == 1` cases first and then continue the loop.

Reply:
Maybe something like this in the loop:

    s1, e1 = w[0]
    if len(w) == 1:
        ...
        continue
    s2, e2 = w[1]
    if e2 > 0:
        ...
    else:
        ...